lib/ui/channel_buffers.dart

```diff
@@ -6,184 +6,334 @@
-  int _dropOverflowItems(int lengthLimit) {
+  /// Removes messages until [length] reaches `lengthLimit`, and returns
+  /// the number of messages removed.
+  int _dropOverflowMessages(int lengthLimit) {
     int result = 0;
     while (_queue.length > lengthLimit) {
-      final T item = _queue.removeFirst();
-      _dropItemCallback?.call(item);
+      final _StoredMessage message = _queue.removeFirst();
+      message.callback(null); // send empty reply to the plugin side
       result += 1;
     }
     return result;
   }
 
-  /// Returns the number of discarded items resulting from resize.
-  int resize(int newSize) {
-    _capacity = newSize;
-    return _dropOverflowItems(newSize);
+  _ChannelCallbackRecord? _channelCallbackRecord;
+
+  /// Sets the listener for this channel.
+  ///
+  /// When there is a listener, messages are sent immediately.
+  ///
+  /// If any messages were queued before the listener is added,
+  /// they are drained asynchronously after this method returns.
+  ///
+  /// Only one listener may be set at a time. Setting a
+  /// new listener clears the previous one.
+  ///
+  /// Callbacks are invoked in their own stack frame and
+  /// use the zone that was current when the callback was
+  /// registered.
+  void setListener(ChannelCallback callback) {
+    final bool needDrain = _channelCallbackRecord == null;
+    _channelCallbackRecord = _ChannelCallbackRecord(callback);
+    if (needDrain && !_draining)
+      _drain();
+  }
+
+  /// Clears the listener for this channel.
+  ///
+  /// When there is no listener, messages are queued, up to [capacity],
+  /// and then discarded in a first-in-first-out fashion.
+  void clearListener() {
+    _channelCallbackRecord = null;
+  }
+
+  /// Drains all the messages in the channel (invoking the currently
+  /// registered listener for each one).
+  ///
+  /// Each message is handled in its own microtask. No messages can
+  /// be queued by plugins while the queue is being drained, but any
+  /// microtasks queued by the handler itself will be processed before
+  /// the next message is handled.
+  ///
+  /// The draining stops if the listener is removed.
+  ///
+  /// See also:
+  ///
+  ///  * [setListener], which is used to register the callback.
+  ///  * [clearListener], which removes it.
+  void _drain() {
+    assert(!_draining);
+    _draining = true;
+    scheduleMicrotask(_drainStep);
   }
-}
 
-/// Signature for [ChannelBuffers.drain].
-typedef DrainChannelCallback = Future<void> Function(ByteData?, PlatformMessageResponseCallback);
+  /// Drains a single message and then reinvokes itself asynchronously.
+  ///
+  /// See [drain] for more details.
+  void _drainStep() {
+    assert(_draining);
+    if (_queue.isNotEmpty && _channelCallbackRecord != null) {
+      final _StoredMessage message = pop();
+      _channelCallbackRecord!.invoke(message.data, message.callback);
+      scheduleMicrotask(_drainStep);
+    } else {
+      _draining = false;
+    }
+  }
+}
```

Comment:
The previous implementation of drain would handle the messages serially, this will potentially parallelize asynchronous operations.  I think that is good since that behavior will match channels when they aren't buffering, correct?

Reply:
The callbacks are synchronous, so they are guaranteed to be complete before the next microtask is scheduled. This matches the way messages are handled outside of drain.

Reply:
Yea, they might spawn asynchronous operations but I think it was erroneous to wait for those operations previously.  This sounds good to me.